src/utfstring.ts

```diff
@@ -0,0 +1,455 @@
+     * @param str The string that is scanned.
+     * @param scanner The scanner that is used to scan the string.
+     * @param charIndex The character index up until which the scan should be performed.
+     * @returns The byte index at which the scan found a match.
+     *          -1 if the scan did not find a match.
+     */
+    private scan(str: string, scanner: RegExp, charIndex: number): number {
+        // optimization: don't iterate unless it's necessary
+        if (!this.containsGraphemeClusterGroup(str)) {
+            return charIndex;
+        }
+
+        let byteIndex = 0;
+        let curCharIndex = 0;
+
+        while (true) {
+            const match = scanner.exec(str);
+            const nextIdx = match ? match.index : str.length;
+
+            while (curCharIndex < charIndex) {
+                if (byteIndex === nextIdx) {
+                    if (curCharIndex < charIndex) {
+                        curCharIndex++;
+
+                        if (match) {
+                            byteIndex += match[0].length;
+                        } else {
+                            byteIndex++;
+                        }
+                    }
+
+                    break;
+                }
+
+                byteIndex++;
+                curCharIndex++;
+            }
+
+            if (curCharIndex === charIndex) {
+                break;
+            } else if (byteIndex >= str.length || !match) {
+                return -1;
+            }
+        }
+
+        return byteIndex;
+    }
+
+    /**
+     * Checks if the given string contains surrogate pairs or regional indicators.
+     * @param str The string that is checked.
+     * @returns True if the given string contains surrogate pairs or regional indicators, false otherwise.
+     */
+    private containsGraphemeClusterGroup(str: string): boolean {
+        return this.graphemeClusterRegex.test(str);
+    }
+
+    /**
+     * Creates a regular expression for scanning strings.
+     * @param extraSources Additional regular expressions that are added to the created regular expression.
+     * @param modifiers Modifier flags for the created regular expression.
+     * @returns The regular expression for scanning string.
+     */
+    private createScanner(extraSources?: string[], modifiers?: string): RegExp {
+        if (extraSources === undefined) {
+            extraSources = ["[^]"];
+        }
+
+        if (modifiers === undefined) {
+            modifiers = "g";
+        }
+
+        let sources = new Array<string>();
+
+        this.graphemeClusterRegexes.forEach((re) => {
+            sources.push(re.source);
+        });
+
+        sources.push(surrogatePairs.source);
+        sources = sources.concat(extraSources);
+
+        return new RegExp(sources.join("|"), modifiers);
+    }
+}
+
+const visual = new UtfStringClass([regionalIndicatorPairs], undefined);
+const UtfString = new UtfStringClass([], visual);
```

Comment:
I think there is currently missing two exports from here https://github.com/camertron/utfstring/blob/master/utfstring.js#L332

Also do you think the export could be done this way instead?

```ts
export const ...
```

Reply:
The export is done just like it was in the former JavaScript version. The `export` object is overwritten a few lines below those `const` assignments with the `UtfString` constant. So import works the same way and existing code doesn't have to be modified at all.